### PR TITLE
feat: Implement printing support in Grain

### PR DESCRIPTION
**Pull request by ospencer**
This PR reimplements some support for printing from JS into Grain. As an added bonus, there is no longer a requirement that types be exported to be able to print them. 🎉 

Included in this PR is a new compiler flag, `--elide-type-information`, which allows you to save a bit of space by not including the information for `print`.

@phated I'm not sure if we want to label this as a breaking change—if anyone depended on the semantics of `toString` for non-exported types this would be a breaking change for them.

### CODE REVIEW

File: runtime/src/core/runner.js [hunk 1]
```diff
@@ -112,22 +53,6 @@ export class GrainRunner {
...
-    view[ptr] = GRAIN_STRING_HEAP_TAG;
-    view[ptr + 1] = v.length;
-    let byteView = this.managedMemory.u8view;
-    for (let i = 0; i < buf.length; ++i) {
-      byteView[i + ptr * 4 + 8] = buf[i];
-    }
-    return userPtr;
-  }
-
-  // [HACK] Temporarily used while we transition to Grain-based runtime
```
Comment (phated) [inline runtime/src/core/runner.js:130]: Why is this comment removed? Does it no longer apply to `grainValueToString`?

File: stdlib/runtime/string.gr [hunk 2]
```diff
@@ -37,6 +33,98 @@ primitive (||) : (Bool, Bool) -> Bool = "@or"
...
+      let fieldName = allocateString(fieldLength)
+      Memory.copy(fieldName + 8n, fields + fieldOffset + 8n, fieldLength)
+      WasmI32.store(fieldArray + (i * 4n), fieldName, 8n)
+
+      fieldOffset += WasmI32.load(fields + fieldOffset, 0n)
+    }
+
+    fieldArray
+  }
+}
```
Comment (phated) [inline stdlib/runtime/string.gr:126]: Are there opportunities for constants with a bunch of these hardcoded offsets and stuff? I forget when we can and can't use them.

File: stdlib/runtime/string.gr [hunk 3]
```diff
@@ -178,11 +263,9 @@ let rec heapValueToString = (ptr, extraIndents, toplevel) => {
       }
     },
     t when t == Tags._GRAIN_RECORD_HEAP_TAG => {
-      // these are tagged ints
-      let moduleId = WasmI32.load(ptr, 4n) >> 1n
-      let typeId = WasmI32.load(ptr, 8n) >> 1n
       let recordArity = WasmI32.load(ptr, 12n)
-      if (!recordTypeExists(moduleId, typeId) || recordArity == 0n) {
+      let fields = getRecordFieldNames(ptr)
+      if (fields == -1n || recordArity == 0n) {
```
Comment (phated) [inline stdlib/runtime/string.gr:268]: What does an arity-0 record actually look like? Why does it exist?

File: stdlib/runtime/string.gr [hunk 4]
```diff
@@ -218,6 +301,8 @@ let rec heapValueToString = (ptr, extraIndents, toplevel) => {
         Memory.decRef(WasmI32.fromGrain(colspace))
         Memory.decRef(WasmI32.fromGrain(comlf))
 
+        Memory.free(fields) // Avoid double-free of record field names
```
Comment (phated) [inline stdlib/runtime/string.gr:304]: Why is this needed now but not before?

**Review by phated — COMMENTED** [review]
So many questions!

File: compiler/src/codegen/compcore.re [hunk 5]
```diff
@@ -2935,9 +2935,151 @@ and compile_instr = (wasm_mod, env, instr) =>
...
+
+  List.iter(
+    meta => {
+      switch (meta) {
+      | ADTMetadata(id, cstrs) =>
+        let section_length =
+          List.fold_left(
+            (total, (_, cstr)) =>
+              total + 12 + round_to_8(String.length(cstr)),
+            8,
```
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/codegen/compcore.re:2990]: They're the sizes of headers for the section. Constants? Maybe. It's dependent on the following code, so it didn't really make sense to make them constants. It might be better to just add a comment that links to the doc I wrote about it.

File: compiler/src/codegen/compcore.re [hunk 6]
```diff
@@ -2935,9 +2935,151 @@ and compile_instr = (wasm_mod, env, instr) =>
...
+            cstrs,
+          );
+        Buffer.add_int32_le(buf, Int32.of_int(section_length));
+        Buffer.add_int32_le(buf, Int32.of_int(id));
+        List.iter(
+          ((id, cstr)) => {
+            let length = String.length(cstr);
+            let aligned_length = round_to_8(length);
+            Buffer.add_int32_le(buf, Int32.of_int(aligned_length + 12));
```
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/codegen/compcore.re:2999]: Nah, just coincidentally the same value.

File: compiler/src/codegen/compcore.re [hunk 7]
```diff
@@ -2935,9 +2935,151 @@ and compile_instr = (wasm_mod, env, instr) =>
...
+            Buffer.add_int32_le(buf, Int32.of_int(id));
+            Buffer.add_int32_le(buf, Int32.of_int(length));
+            Buffer.add_string(buf, cstr);
+            for (_ in 1 to aligned_length - length) {
```
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/codegen/compcore.re:3003]: It does! `to` and `downto` for `for` loops.

File: compiler/src/codegen/compcore.re [hunk 8]
```diff
@@ -2935,9 +2935,151 @@ and compile_instr = (wasm_mod, env, instr) =>
...
+            8,
+            fields,
+          );
+        Buffer.add_int32_le(buf, Int32.of_int(section_length));
+        Buffer.add_int32_le(buf, Int32.of_int(id));
+        List.iter(
+          field => {
+            let length = String.length(field);
+            let aligned_length = round_to_8(length);
+            Buffer.add_int32_le(buf, Int32.of_int(aligned_length + 8));
```
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/codegen/compcore.re:3022]: Same deal.

File: compiler/src/codegen/compcore.re [hunk 9]
```diff
@@ -2935,9 +2935,151 @@ and compile_instr = (wasm_mod, env, instr) =>
...
+            Buffer.add_int32_le(buf, Int32.of_int(length));
+            Buffer.add_string(buf, field);
+            for (_ in 1 to aligned_length - length) {
+              Buffer.add_int8(buf, 0);
+            };
```
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/codegen/compcore.re:3027]: Not a bad idea.

File: compiler/src/codegen/compcore.re [hunk 10]
```diff
@@ -2935,9 +2935,151 @@ and compile_instr = (wasm_mod, env, instr) =>
...
+          },
+          fields,
+        );
+      | ExceptionMetadata(_) => failwith("impossible by partition")
```
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/codegen/compcore.re:3031]: 🤔  When I wrote this I thought the opposite, that I shouldn't bother with making `ExceptionMetadata` since it would just largely be the same as `ADTMetadata`. The partition I'm talking about is the one that happens right above this block of code, and that's just done because there's some preprocessing that needs to happen to them anyway, and after that it's the same as `ADTMetadata`. We could instead do the preprocessing in `linearize.re`, but you'd still need some sort of type to collect the values so you could process them. It also felt like the distinction was better made here, since it's really just that at the code level they're the same, but different elsewhere. I dunno.

File: compiler/src/codegen/compcore.re [hunk 11]
```diff
@@ -3186,9 +3334,16 @@ let compile_globals = (wasm_mod, env, {globals} as prog) => {
 };
 
 let compile_main = (wasm_mod, env, prog) => {
+  let preamble =
+    if (Config.elide_type_info^ || List.length(prog.type_metadata) == 0) {
+      Expression.nop(wasm_mod);
```
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/codegen/compcore.re:3339]: No real reason. I actually changed it to that in https://github.com/grain-lang/grain/pull/563/files. It doesn't make much of a difference because Binaryen takes care of it.

File: runtime/src/core/runner.js (hunk 1 above)
Comment (ospencer) [reply runtime/src/core/runner.js:130]: This particular function was never really a hack... I think the comment was by mistake (since it looks like it was just copy-pasted onto everything).

File: stdlib/runtime/string.gr (hunk 2 above)
Comment (ospencer) [reply stdlib/runtime/string.gr:126]: We can't use them here, though I realized it actually wouldn't be that hard to support it. https://github.com/grain-lang/grain/issues/571

File: stdlib/runtime/string.gr (hunk 3 above)
Comment (ospencer) [reply stdlib/runtime/string.gr:268]: It's currently impossible to make one. I just left this in, but we can delete it. It's also fine as a sanity check. (But probably we can just delete it.)

File: stdlib/runtime/string.gr (hunk 4 above)
Comment (ospencer) [reply stdlib/runtime/string.gr:304]: `fields` is a value that didn't exist until this PR!

File: compiler/src/codegen/compcore.re (hunk 5 above)
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/codegen/compcore.re:2990]: Ah, I already put a comment.

File: compiler/src/codegen/compcore.re (hunk 10 above)
[the remark this answers is not in the document]
Comment (phated) [reply compiler/src/codegen/compcore.re:3033]: Yeah, I understand, but I've found my conceptualization of ADTs to be heavily influenced by the talk "Making Impossible States Impossible" by Richard Feldman

It just feels weird that we have this ADT that we massage and then put impossible cases in our pattern matches.

File: compiler/src/codegen/compcore.re (hunk 11 above)
[the remark this answers is not in the document]
Comment (phated) [reply compiler/src/codegen/compcore.re:3341]: Sounds good!

File: stdlib/runtime/string.gr (hunk 3 above)
Comment (phated) [reply stdlib/runtime/string.gr:268]: Let's delete it! Less code is always nicer

File: compiler/src/codegen/compcore.re (hunk 10 above)
[the remark this answers is not in the document]
Comment (ospencer) [reply compiler/src/codegen/compcore.re:3033]: Yeah. FWIW, the reason why we do that a lot in other places is because we create some types in the parsetree, and then every other stage of the compiler we alias the type instead of creating a new one with just the variants we need. We should probably make an issue to reconsider that.

I can sort of do what you're saying here to get rid of this case.

[the remark this answers is not in the document]
Comment (phated) [reply compiler/src/codegen/compcore.re:3033]: Up to you, I know we do it a lot and it doesn't actually affect the correctness of the compiler. I would be happy with an issue!

**Review by phated — APPROVED** [review]
LGTM with the changes (after tests pass)!! 👍 🎊
